### DISCUSSION TITLE
SettingsControls: SettingsCard + SettingsExpander

### DISCUSSION THREAD

**Discussion by michael-hawker** [CommunityToolkit/Labs-Windows#129] [Experiments]
# Settings Expander

## Problem Statement

The new settings experience in Windows 11 was met with a lot of approval and has quickly been picked up as a common practice in many applications. However, there is no standard control or design template to achieve its behavior, this makes different app's implementations and behaviors inconsistent which is bad for users. It also is more work for developers to redo this effort.

It'd be great to provide a common control that can be used for these scenarios.

## Inspiration

Many community developers have worked on their own implementations of this control/pattern.

## Proposed Solution

It'd be great to have a wrapper around the built-in `Expander` control as a base, and provide a succinct syntax to providing this custom experience, something like (looking at Clock app as example patterns):

### Simple Toggle Button

![image](https://user-images.githubusercontent.com/24302614/169622596-f8f41b96-a99b-48ff-a480-6280e3bca2c5.png)

```xml
<controls:SettingsExpander Icon="{StaticResource ToDoIcon}" Title="To Do" Description="Show tile in the focus session experience">
    <ToggleSwitch IsOn="{x:Bind ViewModel.HasToDo, Mode=TwoWay}"/>
</controls:SettingsExpander>
```

### Single Radio Buttons

Collapsed:
![image](https://user-images.githubusercontent.com/24302614/169622416-aea75b39-b9d4-4b85-b16f-ae64b26dc509.png)

Expanded:
![image](https://user-images.githubusercontent.com/24302614/169622330-89e49c05-0fee-4c96-8928-d061995c9b3c.png)

```xml
<controls:SettingsExpander Icon="{StaticResource PaletteIcon}" Title="App theme" Description="Select which app theme to display">
    <TextBlock Text="{x:Bind SelectedAppThemeText}"/> <!-- Or could be the ToggleSwitch, etc... -->
    <!-- Optional set of items to provide the drop-down without it, it won't expand -->
    <controls:SettingsExpander.Items>
        <controls:SettingsExpanderItem> <!-- Would also support Header in which case content is encapsulated and right-aligned, see next example -->
           <muxc:RadioButtons>
               <x:String>Light</x:String>
               <x:String>Dark</x:String>
               <x:String>Use system setting</x:String>
           </muxc:RadioButtons>
        </controls:SettingsExpanderItem>
    </controls:SettingsExpander.Items>
</controls:SettingsExpander>
```

### Multiple Dropdown Options

Collapsed:
![image](https://user-images.githubusercontent.com/24302614/169622435-4218ad53-0251-4666-8f8b-2c9734687d2a.png)

Expanded:
![image](https://user-images.githubusercontent.com/24302614/169622454-450e8775-0f92-465e-b58d-4c36c88d3ea2.png)

```xml
<controls:SettingsExpander Icon="{StaticResource FocusIcon}" Title="Focus periods" Description="Adjust the length of your focus time, or breaks to fit your needs.">
    <controls:SettingsExpander.Items> <!-- spacing of multiple items with style and corners automatic -->
        <controls:SettingsExpanderItem Header="Focus period">
             <ComboBox .../>
        </controls:SettingsExpanderItem>
        <controls:SettingsExpanderItem Header="Break period">
             <ComboBox .../>
        </controls:SettingsExpanderItem>
    </controls:SettingsExpander.Items>
</controls:SettingsExpander>
```
### Adaptive layout
There should be logic to support smaller screen sizes:

![AdaptiveSettingsCard](https://user-images.githubusercontent.com/9866362/188679317-3853add7-8397-4687-a3dd-8fc93c517cbf.gif)

## API Reference Docs
- [Expander](https://docs.microsoft.com/windows/winui/api/microsoft.ui.xaml.controls.expander)

## References
- https://github.com/Richasy/ExpanderEx
- https://github.com/microsoft/PowerToys/tree/main/src/settings-ui/Settings.UI/Controls/Setting
- https://github.com/ghost1372/SettingsUI

**Comment by niels9001** [comment]
Nice @michael-hawker! 

Here are some thoughts on what things it should support:


**Main card properties**
| Property name  | Type | Comments |
| ------------- | ------------- | ------------- |
| Header  | string  |
| Description  | object | This would allow us to also support things like HyperLinkButtons (which is common across Settings)
| Icon | IconElement | By default it should use ThemeSymbolFontFamily, but this would allow developers to also set images as icons
| Content | object | This can be anything, empty by default. Allows users to also add multiple controls (with e.g. a horizontal StackPanel)
| Items | SettingsExpanderItem | 

**SettingsExpanderItem**
This should be fully customizable - it could host a list of CheckBoxes, but also entirely custom UIs / panels. And a mix of those things :)


**Visual style**
Currently, there are many different implementations of this control across various inbox apps with all slightly different styles (e.g. margins, sizes, colors). I'd suggest that the implementation in the W11 Settings app is our northstar - I can check in with the Design team to get the correct brushes and margins.


**Accesibility**
![image](https://user-images.githubusercontent.com/9866362/170039194-e1034a82-ce5d-42f6-9747-43c00a9933ec.png)

Narrator should annouce this as Header + Content control. For example: "Windows display language - English (United States".
Additionally, the Description can be set as a HelpText (or by default?) and the Expanded state should be called out.


================

**SettingsButton**
W11 Settings also provides buttons to e.g. navigate to a subpage. I'm guessing this should be a separate control right? (If any.. since it's basically a Button with a specific layout - but would be nice to have). Would love to know your thoughts on that.
![image](https://user-images.githubusercontent.com/9866362/170040085-504f4e48-893e-4932-af66-f16927246f61.png)

**Reply by michael-hawker** [reply]
Thanks @niels9001!

For the `SettingsButton` scenario, I could see it just using this control as well still, with an `Alignment` property to help:

```xml
<controls:SettingsExpander Icon="{StaticResource MonitorIcon}" Title="Display" Description="Monitors, brightness, night light, display profle">
    <Button HorizontalAlignment="Right" Command="{x:Bind OpenOtherPageCommand}" Content="&gt;"/>
</controls:SettingsExpander>
```

Since they should be able to take any content, it should be straight-forward. Though I guess in this case the hit box is just over the button. Where as in the Settings experience, the whole card itself is the button, eh? 🤔

It's frustrating that there's no way to cancel Expanding of an Expander... https://github.com/microsoft/microsoft-ui-xaml/issues/4407 - as then we could intercept that and use it as a button in this case.

I suppose in either case, we'll probably need a `StyleSelector` to begin with, which could have three modes:

- If it has `Items` it's an Expander as a base
- If it has `IsClickEnabled` (or some other name) makes it a Button
- Otherwise it's just a static 'card', as there'll be some other content/control inside like a ToggleSwitch or ComboBox.

Then it'd just be:

```xml
<controls:SettingsExpander Icon="{StaticResource MonitorIcon}" Title="Display" Description="Monitors, brightness, night light, display profle"
                           IsClickEnabled="True" Click="SomeEventOrHookWithBehavior"/>
```

Though if we add `Click` not sure if we want to directly add `Command` as well or leverage behaviors for that. Not sure how common this scenario will be over just putting direct Items or other controls inside?

**Reply by niels9001** [reply]
> For the `SettingsButton` scenario, I could see it just using this control as well still, with an `Alignment` property to help:

> Since they should be able to take any content, it should be straight-forward. Though I guess in this case the hit box is just over the button. Where as in the Settings experience, the whole card itself is the button, eh? 🤔

Yep, and additional hover states when you mouse over etc.

> I suppose in either case, we'll probably need a `StyleSelector` to begin with, which could have three modes:
> 
> * If it has `Items` it's an Expander as a base
> * If it has `IsClickEnabled` (or some other name) makes it a Button
> * Otherwise it's just a static 'card', as there'll be some other content/control inside like a ToggleSwitch or ComboBox.

Yeah, that would be nice. If it would make things overly complex, I could imagine 2 separate controls (SettingsExpander and SettingsCard) could work (with the latter having the IsClickEnabled property that will adopt the Button style).

**Reply by michael-hawker** [reply]
I've gotten a bit crafty with selectors these days (see `TokenizingTextBox` where we do wonky things too)...

I don't think it'd be as crazy here and we could investigate what it'd look like. If it's a single control it makes it nice as you could use an `ItemsControl` and a DataTemplate to bind properties to a the `SettingsExpander` and potentially make it a bit more data driven. Though that starts to fall apart a bit if you want more complex options and choices for that data...

**Reply by michael-hawker** [reply]
Talked to team, sounds like we may want a `SettingsCard` for the flat and button scenarios, and then the `SettingsExpander` for the more complex scenarios.

Some more examples which I think would fit in that framework are here as well:

![image](https://user-images.githubusercontent.com/24302614/172440509-4355361d-05ed-468f-a78b-d235d912c8bc.png)

![image](https://user-images.githubusercontent.com/24302614/172441514-6afd62a3-236a-4b37-8269-cc36c3567914.png)

![image](https://user-images.githubusercontent.com/24302614/172441600-fcbe35ac-e5ca-4c9e-b334-175290426750.png)

**Comment by niels9001** [comment]
After our recent discussion, here are some updated specifications.

It seems to make most sense to create 2 separate controls: **SettingsCard** and **SettingsExpander**.

# SettingsCard
A single card that has properties to set the icon, header, description and content (e.g. a ToggleSwitch), right aligned. It provides developers a way to display a single setting.

![image](https://user-images.githubusercontent.com/9866362/173779248-45a48aa5-3613-4c73-8eba-6c394ecdba1d.png)


```xml
<controls:SettingsCard Icon="&#xE08F;" Title="This is the header" Description="And this is the description">
    <ToggleSwitch IsOn="{x:Bind ViewModel.HasToDo, Mode=TwoWay}"/>
</controls:SettingsCard>
```

| Property name  | Type | Comments |
| ------------- | ------------- | ------------- |
| Icon | IconElement | By default it should use ThemeSymbolFontFamily, but this would allow developers to also set images as icons.
| Header  | string  |
| Description  | object | Should be a TextBlock by default, but can also be customized by non-string objects (e.g. to display a HyperLinkButton)
| Content | object | This can be anything, empty by default. Allows users to also add multiple controls (with e.g. a horizontal StackPanel)
| _IsItemClickEnabled_ | _bool_ | _If true, this control turns into a Button_


</br>
</br>
</br>

---


# SettingsExpander
A control that enherites from WinUI.Expander thas properties to set the icon, header, description and content. It provides developers a way to group multiple options into a single UI control.

Automatic layouting of controls:

![image](https://user-images.githubusercontent.com/9866362/173788115-8fcf2f36-c323-43f8-b545-2f011e5cb1fe.png)

```xml
<controls:SettingsExpander Icon="{StaticResource FocusIcon}" Title="This is the header" Description="And this is the description">
    <controls:SettingsExpander.Items> <!-- spacing of multiple items with style and corners automatic -->
        <controls:SettingsExpanderItem Header="This is a first option">
             <ComboBox .../>
        </controls:SettingsExpanderItem>
        <controls:SettingsExpanderItem Header="And a second option">
             <ComboBox .../>
        </controls:SettingsExpanderItem>
    <controls:SettingsExpanderItem Header="And a third option">
             <Button.../>
        </controls:SettingsExpanderItem>
    </controls:SettingsExpander.Items>
</controls:SettingsExpander>
```

Hosting a custom panel:

![image](https://user-images.githubusercontent.com/9866362/173789541-9a42a022-29a3-4032-a282-752a8a4bd1f2.png)

```xml
<controls:SettingsExpander Icon="{StaticResource FocusIcon}" Title="This is the header" Description="And this is the description">
    <controls:SettingsExpander.Items>
        <controls:SettingsExpanderItem Header="This is a first option">
             <StackPanel ...>
                  <TextBlock .../>
                   <Button .../>
                   <GridView .../>
              <StackPanel/>
        </controls:SettingsExpanderItem>
    </controls:SettingsExpander.Items>
</controls:SettingsExpander>
```

| Property name  | Type | Comments |
| ------------- | ------------- | ------------- |
| Icon | IconElement | By default it should use ThemeSymbolFontFamily, but this would allow developers to also set images as icons.
| Header  | string  |
| Description  | object | Should be a TextBlock by default, but can also be customized by non-string objects (e.g. to display a HyperLinkButton)
| Items | SettingsExpanderItem | This should be flexible as this could host custom layouts (e.g. a Grid) but also a single option (that would require automatic styling, based on the control type. A ComboBox would be right aligned, a CheckBox left etc..


### Open questions
- SettingsExpanderItem - ?

**Reply by michael-hawker** [reply]
Awesome @niels9001. The the card, should the property be `IsClickEnabled` as it's not an "item" itself like a traditional `ListView`?

**Reply by michael-hawker** [reply]
Should `Title` be `Header`?

In either case Title/Description should be objects and use ContentPresenters to accept more than just strings (like Button content). That way someone could add a Hyperlink for the description to click on.

**Comment by Jay-o-Way** [comment]
Question: in the first example of the settings expander I see a few settings separated by a line. What element is used for the line?

**Reply by michael-hawker** [reply]
Can you clarify which image you're referring to? The line is automatic between SettingsExpanderItems.

**Reply by Jay-o-Way** [reply]
The one with "first, second, third". I suspect it should be a MenuFlyoutSeparator?

**Reply by niels9001** [reply]
@Jay-o-Way We set the Border of the root grid to 0,1,0,0 (SettingsExpanderItemBorderThickness) so it correctly renders for every SettingsExpanderMenuItem. The XAML is part of this PR: https://github.com/CommunityToolkit/Labs-Windows/pull/253

**Comment by 0x5bfa** [comment]
When adding these settings control in code-behind, it should be easy to create a title for the group as well. This can be seen in PowerToy's settings control

XAML code structure is here:

```xaml
<uc:SettingsGroup Title="Input">
    <uc:SettingsGroup.Items>
        <uc:SettingsCard Icon="" Title="" Description="">
            <Button Background="transparent" BorderThickness="0">
                <FontIcon Glyph="" />
            <Button>
        </uc:SettingsCard>
        <uc:SettingsCard Icon="" Title="" Description="">
            <Button Background="transparent" BorderThickness="0">
                <FontIcon Glyph="" />
            <Button>
        </uc:SettingsCard>
    </uc:SettingsGroup.Items>
</uc:SettingsGroup>
```
![image](https://user-images.githubusercontent.com/62196528/213224948-17606f3d-a577-4d7e-9a50-25dab9f643bb.png)

Also, FYI, in Windows 11 Settings, SettingsExpander has footer border(background is not same as above items):

![image](https://user-images.githubusercontent.com/62196528/213228110-bdb8427f-7068-4bf2-9e2c-80ff730c36ef.png)

**Reply by michael-hawker** [reply]
@niels9001 we should probably just include these values in the control template? (Though I know WinUI makes it hard to use resources from controls due to resource lookup order.)

https://github.com/CommunityToolkit/Labs-Windows/blob/bfb745f03d820bb1529f5f6c967a56314fb9d50d/labs/SettingsControls/samples/SettingsControls.Samples/SettingsPageExample.xaml#L10-L24

@onein528 we didn't want to have a control just to group settings as that's very rigid, we just provide the complete sample for design reference at the moment. 

We plan to look at a more general [form control](https://github.com/microsoft/microsoft-ui-xaml/issues/82) experiment to help with this and other scenarios for grouping components on a page with headers in a specific layout. We just haven't gotten to that yet. Should probably open a new discussion/proposal here for tracking to copy over the relevant bits from that issue and the related possible panel PR here that sparked that discussion: https://github.com/CommunityToolkit/WindowsCommunityToolkit/pull/3381

**Comment by michael-hawker** [comment]
This was shipped in the [8.0 Release](https://github.com/CommunityToolkit/Windows/releases/tag/v8.0.230907), please update to use that version of the component. Any further issues/requests should start in the main repository.
